R/data-combined.R

```diff
@@ -0,0 +1,867 @@
+        simulationResults = simulationResults,
+        quantitiesOrPaths = quantitiesOrPaths,
+        population        = population,
+        individualIds     = individualIds
+      )
+
+      # if names are not present, use paths as unique names
+      if (!"name" %in% names(data)) {
+        data <- dplyr::mutate(data, name = paths)
+      }
+
+      # if alternative names are provided, use them
+      if (!is.null(names)) {
+        data <- private$.renameDatasets(data, names) %>%
+          dplyr::mutate(
+            name = dplyr::case_when(
+              is.na(name) ~ paths,
+              TRUE ~ name
+            )
+          )
+      }
+
+      # add column with grouping information and then
+      #
+      # add a column describing the type of data;
+      # simulation results never have errors, but this leads to inconsistent
+      # output, so also add a column for `yErrorValues` outputs and then
+      #
+      # rename according to column naming conventions for `DataSet` and then
+      data <- data %>%
+        private$.addGroupCol(groups) %>%
+        dplyr::mutate(
+          dataType     = "simulated",
+          yErrorValues = NA_real_
+        ) %>%
+        dplyr::rename(
+          "xValues"    = "Time",
+          "xUnit"      = "TimeUnit",
+          "xDimension" = "TimeDimension",
+          "yValues"    = "simulationValues",
+          "yUnit"      = "unit",
+          "yDimension" = "dimension"
+        )
+
+      return(data)
+    },
+
+    ## dataframe modifiers ---------------------
+
+    # Add a new group column
+    #
+    # If no grouping is specified for a dataset, the group will be `NA`.
+    #
+    # While visualizing such data, the datset name can instead be used as a
+    # grouping variable, but this association will happen in the plotting
+    # function itself, and not in the output produced by the current object.
+
+    .addGroupCol = function(data, groups = NULL) {
+      if (!is.null(groups)) {
+        data <- data %>%
+          dplyr::group_by(name) %>%
+          tidyr::nest() %>%
+          dplyr::ungroup() %>%
+          dplyr::mutate(group = groups) %>%
+          tidyr::unnest(cols = c(data))
+      } else {
+        data <- data %>% dplyr::mutate(group = NA_character_)
+      }
+
+      return(data)
+    },
+
+    # update the combined dataframe in place
+
+    .updateDF = function(dataCurrent = NULL, dataNew = NULL) {
+      # if there is already data, add new data at the bottom
+      # since the `toDataFrame()` method arranges the data by dataset name,
+      # this will not introduce any order effects, i.e., it doesn't matter
+      # if the user entered `DataSet` objects first and then `SimulationResults`
+      # objects, or vice versa
+      if (!is.null(dataCurrent) && !is.null(dataNew)) {
+        # check for duplicated datasets between the current and the new dataframes
+        dupDatasets <- intersect(unique(dataCurrent$name), unique(dataNew$name))
```

Comment:
I don't understand this. How can we have duplicate?

Reply:
Sorry, poor phrasing on my part. I have changed this to the following:

```r
        # check if the new dataset(s) entered are already present in the
        # internal combined dataframe
        dupDatasets <- intersect(unique(dataCurrent$name), unique(dataNew$name))

        # if this is the case, then replace the older datasets with the newer
        # versions of the same datasets
        #
        # e.g. someone can all `$addSimulationResults(dataSet1)` and then again
        # call `$addSimulationResults(dataSet1)` with the same class instance.
        # In this case, dataframe created in the latter call will replace the
        # one created in the former call
        if (length(dupDatasets) > 0L) {
          dataCurrent <- dplyr::filter(dataCurrent, !name %in% dupDatasets)
        }
```

Reply:
I am punting this to @PavelBal, who wanted this scenario to be supported.

Reply:
How about to cover the following context:

I run a simulation and call `myCombDat$addSimulationResults(simResults1)`.

But then I realized I did something wrong while running the simulation, so I rerun it. In this case, it will be convenient to just do the same call again with the assurance that the new results will replace the old ones.

Does that make sense? I am guessing this as a non-user. Maybe Pavel has a much more compelling case to make.

Reply:
sure